components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/FxaActionProcessor.kt

```diff
@@ -0,0 +1,342 @@
+    private suspend fun<T> withNetworkRetry(operation: suspend () -> T): T {
+        while (true) {
+            try {
+                if (simulateNetworkErrorFlag) {
+                    simulateNetworkErrorFlag = false
+                    throw FxaException.Network("Simulated Error")
+                }
+                return operation()
+            } catch (e: FxaException.Network) {
+                if (retryLogic.shouldRetryAfterNetworkError()) {
+                    Log.d(LOG_TAG, "Network error: retrying")
+                    continue
+                } else {
+                    Log.d(LOG_TAG, "Network error: not retrying")
+                    throw e
+                }
+            }
+        }
+    }
+
+    // Perform an operation, retrying after network errors and calling checkAuthorizationStatus
+    // after auth errors
+    private suspend fun<T> withRetry(operation: suspend () -> T): T {
+        while (true) {
+            try {
+                return withNetworkRetry(operation)
+            } catch (e: FxaException.Authentication) {
+                val currentState = currentState
+
+                if (currentState !is FxaAuthState.Connected) {
+                    throw e
+                }
+
+                if (retryLogic.shouldRecheckAuthStatus()) {
+                    Log.d(LOG_TAG, "Auth error: re-checking")
+                    handleCheckAuthorization(currentState)
+                } else {
+                    Log.d(LOG_TAG, "Auth error: disconnecting")
+                    inner.disconnectFromAuthIssues()
+                    persistState()
+                    changeState(FxaAuthState.Disconnected(true), FxaAuthStateTransition.AUTH_CHECK_FAILED)
+                }
+
+                if (this.currentState is FxaAuthState.Connected) {
+                    continue
+                } else {
+                    throw e
+                }
+            }
+        }
+    }
+
+    private suspend fun handleBeginOAuthFlow(currentState: FxaAuthState.Disconnected, action: FxaAction.BeginOAuthFlow) {
+        handleBeginEitherOAuthFlow(currentState, action.result) {
+            inner.beginOauthFlow(action.scopes.toList(), action.entrypoint, MetricsParams(mapOf()))
+        }
+    }
+
+    private suspend fun handleBeginPairingFlow(currentState: FxaAuthState.Disconnected, action: FxaAction.BeginPairingFlow) {
+        handleBeginEitherOAuthFlow(currentState, action.result) {
+            inner.beginPairingFlow(action.pairingUrl, action.scopes.toList(), action.entrypoint, MetricsParams(mapOf()))
+        }
+    }
+
+    private suspend fun handleBeginEitherOAuthFlow(currentState: FxaAuthState.Disconnected, result: CompletableDeferred<String?>?, operation: () -> String) {
+        try {
+            val url = withRetry { operation() }
+            persistState()
+            changeState(currentState.copy(connecting = true), FxaAuthStateTransition.OAUTH_STARTED)
+            sendEvent(FxaEvent.BeginOAuthFlow(url))
```

Comment:
Good point, I'll add an extra param here.  If a consumer doesn't care than they can just ignore it.

Reply:
Actually, I'm not so sure yet.  I still think this is a good point, but I think it applies more to the `FxaEvent.AuthEvent` (formally `StateChanged`).  I was considering adding field for the OAuth flow kind there, but it's not so easy since there can be multiple flows.

Firefox android doesn't need this yet, so I think we should wait on implementing it, maybe alongside a requirement that only one OAuth flow is in progress at any time.